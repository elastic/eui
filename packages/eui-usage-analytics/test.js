/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const Codeowners = require('codeowners');
const temp = new Codeowners("../../../kibana");

const { scan } = require('./scan');

const runScan = async () => {
  const scanResult = await scan();
  console.log(scanResult);
};

runScan();
