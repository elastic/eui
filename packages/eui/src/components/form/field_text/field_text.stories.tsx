/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import {
  disableStorybookControls,
  moveStorybookControlsToCategory,
} from '../../../../.storybook/utils';

import { EuiFieldText, EuiFieldTextProps } from './field_text';

const meta: Meta<EuiFieldTextProps> = {
  title: 'Forms/EuiFieldText',
  component: EuiFieldText,
  argTypes: {
    // For quicker/easier QA
    icon: { control: 'text' },
    prepend: { control: 'text' },
    append: { control: 'text' },
    value: { control: 'text' },
  },
  args: {
    // Component defaults
    compressed: false,
    fullWidth: false,
    isInvalid: false,
    isLoading: false,
    disabled: false,
    readOnly: false,
    controlOnly: false,
    // Added for easier testing
    placeholder: 'EuiFieldText',
    id: '',
    name: '',
  },
};

export default meta;
type Story = StoryObj<EuiFieldTextProps>;
disableStorybookControls(meta, ['inputRef']);

export const Playground: Story = {};

export const IconShape: Story = {
  parameters: {
    controls: {
      include: [
        'icon',
        'compressed',
        'fullWidth',
        'prepend',
        'append',
        'isInvalid',
        'isLoading',
        'disabled',
        'readOnly',
      ],
    },
  },
  argTypes: { icon: { control: 'object' } },
  args: { icon: { type: 'warning', color: 'warning', side: 'left' } },
};
// Move other props below the icon prop
moveStorybookControlsToCategory(IconShape, [
  'compressed',
  'fullWidth',
  'isInvalid',
  'isLoading',
  'disabled',
  'readOnly',
  'prepend',
  'append',
]);

export const AutoFill: Story = {
  parameters: {
    controls: { include: ['name', 'isInvalid'] },
    loki: { skip: true },
  },
  decorators: [
    (Story) => (
      <form action="#">
        In Chrome: Type any text, press Enter, then go back and select the
        autofill suggestion. Test light+dark mode as well as invalid state
        <br />
        <br />
        <Story />
      </form>
    ),
  ],
  args: {
    name: 'autofill-test',
  },
};
