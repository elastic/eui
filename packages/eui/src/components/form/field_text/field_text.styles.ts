/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { css } from '@emotion/react';

import { UseEuiTheme } from '../../../services';
import { euiFormControlStyles } from '../form.styles';

export const euiFieldTextStyles = (euiThemeContext: UseEuiTheme) => {
  const formStyles = euiFormControlStyles(euiThemeContext);

  return {
    euiFieldText: css`
      ${formStyles.shared}

      &:invalid {
        ${formStyles.invalid}
      }

      &:focus {
        ${formStyles.focus}
      }

      &[readOnly] {
        ${formStyles.readOnly}
      }

      &:disabled {
        ${formStyles.disabled}
      }

      &:autofill {
        ${formStyles.autoFill}
      }
    `,

    // Skip the css() on the default height to avoid generating a className
    uncompressed: formStyles.uncompressed,
    compressed: css(formStyles.compressed),

    // Skip the css() on the default width to avoid generating a className
    formWidth: formStyles.formWidth,
    fullWidth: css(formStyles.fullWidth),

    // Layout modifiers
    inGroup: css(formStyles.inGroup),
    controlOnly: css`
      .euiFormControlLayout--group & {
        ${formStyles.inGroup}
      }
    `,
  };
};
