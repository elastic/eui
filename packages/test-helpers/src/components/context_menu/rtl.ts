/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { fireEvent, waitFor } from '@testing-library/react';

const waitForPanelTransition = async () => {
  // Used document instead of container or screen due to context menus living in portals
  const getPanels = () => document.querySelectorAll('.euiContextMenuPanel');

  // 2 panels will appear for the transition animation
  await waitFor(() => {
    expect(getPanels().length).toEqual(2);
  });

  // Outgoing panel will be removed on animation end
  fireEvent.animationEnd(getPanels()[0]);
  if (getPanels().length > 1) {
    fireEvent.animationEnd(getPanels()[1]);
  }

  // Transition/animation is done once we're back to 1 panel
  expect(getPanels().length).toEqual(1);
};

export const EuiContextMenuTestHelpers = {
  waitForPanelTransition,
};
