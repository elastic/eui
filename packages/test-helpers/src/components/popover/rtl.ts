/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { waitFor } from '@testing-library/react';

const waitForOpen = () => {
  return waitFor(() => {
    const openPopover = document.querySelector('[data-popover-open]');
    expect(openPopover).toBeTruthy();
  });
}

const waitForClose = () => {
  return waitFor(() => {
    const openPopover = document.querySelector('[data-popover-open]');
    expect(openPopover).toBeFalsy();
  });
};

export const EuiPopoverTestHelpers = {
  waitForOpen,
  waitForClose,
};
