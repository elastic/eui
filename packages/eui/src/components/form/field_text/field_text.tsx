/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { InputHTMLAttributes, Ref, FunctionComponent } from 'react';
import classNames from 'classnames';

import { useEuiMemoizedStyles } from '../../../services';
import { CommonProps } from '../../common';
import {
  EuiFormControlLayout,
  EuiFormControlLayoutProps,
} from '../form_control_layout';
import { EuiValidatableControl } from '../validatable_control';
import { useFormContext } from '../eui_form_context';

import { euiFieldTextStyles } from './field_text.styles';

export type EuiFieldTextProps = InputHTMLAttributes<HTMLInputElement> &
  CommonProps & {
    icon?: EuiFormControlLayoutProps['icon'];
    isInvalid?: boolean;
    /**
     * Expand to fill 100% of the parent.
     * Defaults to `fullWidth` prop of `<EuiForm>`.
     * @default false
     */
    fullWidth?: boolean;
    isLoading?: boolean;
    /**
     * Prevents user from changing input.
     *
     * Defaults to the value of `disabled` unless explicity defined otherwise.
     */
    readOnly?: boolean;
    inputRef?: Ref<HTMLInputElement>;

    /**
     * Creates an input group with element(s) coming before input.
     * `string` | `ReactElement` or an array of these
     */
    prepend?: EuiFormControlLayoutProps['prepend'];

    /**
     * Creates an input group with element(s) coming after input.
     * `string` | `ReactElement` or an array of these
     */
    append?: EuiFormControlLayoutProps['append'];

    /**
     * Completely removes form control layout wrapper and ignores
     * icon, prepend, and append. Best used inside EuiFormControlLayoutDelimited.
     */
    controlOnly?: boolean;

    /**
     * when `true` creates a shorter height input
     */
    compressed?: boolean;
  };

export const EuiFieldText: FunctionComponent<EuiFieldTextProps> = (props) => {
  const { defaultFullWidth } = useFormContext();
  const {
    id,
    name,
    placeholder,
    value,
    className,
    icon,
    isInvalid,
    inputRef,
    fullWidth = defaultFullWidth,
    isLoading,
    compressed,
    prepend,
    append,
    disabled,
    readOnly = disabled, // sync to prevent onChange unless explicitly defined
    controlOnly,
    ...rest
  } = props;

  const classes = classNames('euiFieldText', className, {
    'euiFieldText-isLoading': isLoading,
  });

  const styles = useEuiMemoizedStyles(euiFieldTextStyles);
  const cssStyles = [
    styles.euiFieldText,
    compressed ? styles.compressed : styles.uncompressed,
    fullWidth ? styles.fullWidth : styles.formWidth,
    !controlOnly && (prepend || append) && styles.inGroup,
    controlOnly && styles.controlOnly,
  ];

  const control = (
    <EuiValidatableControl isInvalid={isInvalid}>
      <input
        type="text"
        id={id}
        name={name}
        placeholder={placeholder}
        className={classes}
        css={cssStyles}
        value={value}
        ref={inputRef}
        disabled={disabled}
        readOnly={readOnly}
        {...rest}
      />
    </EuiValidatableControl>
  );

  if (controlOnly) return control;

  return (
    <EuiFormControlLayout
      icon={icon}
      fullWidth={fullWidth}
      isLoading={isLoading}
      isInvalid={isInvalid}
      compressed={compressed}
      isDisabled={disabled}
      readOnly={readOnly}
      prepend={prepend}
      append={append}
      inputId={id}
    >
      {control}
    </EuiFormControlLayout>
  );
};
