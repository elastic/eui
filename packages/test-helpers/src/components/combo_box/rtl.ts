/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { fireEvent, waitFor } from '@testing-library/react';
import { screen } from '@testing-library/react';
import { EuiPopoverTestHelpers } from '../popover/rtl';

const showOptions = async () => {
  fireEvent.click(screen.getByTestSubject('comboBoxToggleListButton'));
  await EuiPopoverTestHelpers.waitForOpen();
  await waitFor(() => {
    expect(screen.getByRole('listbox')).toBeInTheDocument();
  });
};

export const EuiComboBoxTestHelpers = {
  showOptions,
};
