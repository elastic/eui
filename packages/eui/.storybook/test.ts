/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { queries, within, waitFor, fireEvent, expect } from '@storybook/test';
import * as dataTestSubjQueries from '../src/test/rtl/data_test_subj_queries';

/**
 * Custom Storybook within util with EUI query helpers
 * + additional chained async/waitFor component utils
 *
 * @see https://storybook.js.org/docs/writing-stories/play-function#writing-stories-with-the-play-function
 * @see https://testing-library.com/docs/dom-testing-library/api-within/
 */
const customWithin = (canvasElement: HTMLElement) => {
  const canvas = within<typeof queries & typeof dataTestSubjQueries>(
    canvasElement,
    { ...queries, ...dataTestSubjQueries }
  );

  return {
    ...canvas,

    /**
     * 1. Loki doesn't like userEvent, only fireEvent
     * 2. Storybook fires fireEvents too early (esp. on page load), so we add a waitFor
     */
    waitForAndClick: async (testSubject: string) => {
      await waitFor(() =>
        expect(canvas.getByTestSubject(testSubject)).toBeInTheDocument()
      );
      await fireEvent.click(canvas.getByTestSubject(testSubject));
    },

    waitForEuiPopoverVisible: async () =>
      await waitFor(() =>
        expect(canvasElement.querySelector('[data-popover-open]')).toBeVisible()
      ),
    waitForEuiPopoverHidden: async () =>
      await waitFor(() =>
        expect(
          canvasElement.querySelector('[data-popover-panel]')
        ).not.toBeInTheDocument()
      ),
  };
};

export { customWithin as within };
