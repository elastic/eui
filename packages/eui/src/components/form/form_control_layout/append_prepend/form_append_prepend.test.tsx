/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { fireEvent } from '@testing-library/react';

import { shouldRenderCustomStyles } from '../../../../test/internal';
import {
  EuiFormAppend,
  EuiFormAppendPrepend,
  EuiFormAppendPrependProps,
  EuiFormPrepend,
} from './form_append_prepend';
import { requiredProps } from '../../../../test';
import { render } from '../../../../test/rtl';
import { EuiFieldText } from '../../field_text';

const sharedProps = {
  label: 'Label',
  'data-test-subj': 'euiFormAppendPrepend',
};

const defaultProps = {
  ...sharedProps,
  side: 'append' as EuiFormAppendPrependProps['side'],
};

describe('EuiFormAppend', () => {
  shouldRenderCustomStyles(<EuiFormAppend />);

  it('is rendered', () => {
    const { container, getByTestSubject } = render(
      <EuiFormAppend {...sharedProps} {...requiredProps} />
    );

    const classes = Object.values(
      getByTestSubject(requiredProps['data-test-subj']).classList
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(classes[0]).toBe('euiFormAppend');
  });
});

describe('EuiFormPrepend', () => {
  shouldRenderCustomStyles(<EuiFormPrepend />);

  it('is rendered', () => {
    const { container, getByTestSubject } = render(
      <EuiFormPrepend {...defaultProps} {...requiredProps} />
    );

    const classes = Object.values(
      getByTestSubject(requiredProps['data-test-subj']).classList
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(classes[0]).toBe('euiFormPrepend');
  });
});

describe('EuiFormAppendPrepend', () => {
  shouldRenderCustomStyles(<EuiFormAppend />);

  it('is rendered', () => {
    const { container } = render(
      <EuiFormAppend {...defaultProps} {...requiredProps} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('props', () => {
    describe('element', () => {
      it('renders a div', () => {
        const { getByTestSubject } = render(
          <EuiFormAppendPrepend {...defaultProps} element="div" />
        );

        const element = getByTestSubject(defaultProps['data-test-subj']);

        expect(element).toBeInTheDocument();
        expect(element.tagName).toBe('DIV');
      });

      it('renders a button', () => {
        const { getByTestSubject } = render(
          <EuiFormAppendPrepend {...defaultProps} element="button" />
        );

        const element = getByTestSubject(defaultProps['data-test-subj']);

        expect(element).toBeInTheDocument();
        expect(element.tagName).toBe('BUTTON');
      });
    });

    describe('label', () => {
      it('renders a label element', () => {
        const { getByText } = render(
          <EuiFormAppendPrepend {...defaultProps} />
        );

        const element = getByText(defaultProps.label);

        expect(element).toBeInTheDocument();
        expect(element.tagName).toBe('LABEL');
      });

      it('renders a span element for buttons', () => {
        const { getByText } = render(
          <EuiFormAppendPrepend {...defaultProps} element="button" />
        );

        const element = getByText(defaultProps.label);

        expect(element).toBeInTheDocument();
        expect(element.tagName).toBe('SPAN');
      });
    });

    describe('iconLeft', () => {
      it('renders an icon on the left side', () => {
        const { getByTestSubject } = render(
          <EuiFormAppendPrepend {...defaultProps} iconLeft="faceHappy" />
        );

        expect(
          getByTestSubject(defaultProps['data-test-subj']).firstChild
        ).toHaveAttribute('data-euiicon-type', 'faceHappy');
      });
    });

    describe('iconRight', () => {
      it('renders an icon on the left side', () => {
        const { getByTestSubject } = render(
          <EuiFormAppendPrepend {...defaultProps} iconRight="faceHappy" />
        );

        expect(
          getByTestSubject(defaultProps['data-test-subj']).lastChild
        ).toHaveAttribute('data-euiicon-type', 'faceHappy');
      });
    });

    describe('children', () => {
      it('renders', () => {
        const { side, 'data-test-subj': dataTestSubj } = defaultProps;

        const { getByTestSubject } = render(
          <EuiFormAppendPrepend side={side} data-test-subj={dataTestSubj}>
            <span>Content</span>
          </EuiFormAppendPrepend>
        );

        const element = getByTestSubject(dataTestSubj);

        expect(element.children.length).toBe(1);
        expect(element.firstChild).toHaveTextContent('Content');
      });

      it('renders `children` as last child', () => {
        const { getByTestSubject } = render(
          <EuiFormAppendPrepend {...defaultProps}>
            <span>Content</span>
          </EuiFormAppendPrepend>
        );

        const element = getByTestSubject(defaultProps['data-test-subj']);

        expect(element.firstChild).toHaveTextContent(defaultProps.label);
        expect(element.lastChild).toHaveTextContent('Content');
      });
    });

    describe('inputId', () => {
      it('renders `for` attribute when `inputId` is passed', () => {
        const { getByText } = render(
          <EuiFormAppendPrepend {...defaultProps} inputId="testId" />
        );

        expect(getByText(defaultProps.label)).toHaveAttribute('for', 'testId');
      });

      it('does not render `for` attribute for buttons when `inputId` is passed', () => {
        const { getByText } = render(
          <EuiFormAppendPrepend
            {...defaultProps}
            inputId="testId"
            element="button"
          />
        );

        expect(getByText(defaultProps.label)).not.toHaveAttribute('for');
      });

      it('renders `for` attribute when `id` is set on the parent form element', () => {
        const { getByText } = render(
          <EuiFieldText
            id="testId"
            append={<EuiFormAppendPrepend {...defaultProps} />}
          />
        );

        expect(getByText(defaultProps.label)).toHaveAttribute('for', 'testId');
      });
    });

    describe('compressed', () => {
      it('renders compressed styles', () => {
        const { getByTestSubject } = render(
          <EuiFormAppendPrepend {...defaultProps} compressed />
        );

        const classes = Object.values(
          getByTestSubject(defaultProps['data-test-subj']).classList
        );

        expect(classes.some((clx) => clx.includes('compressed'))).toBe(true);
      });

      it('renders compressed styles when the parent form element is compressed', () => {
        const { getByTestSubject } = render(
          <EuiFieldText
            compressed
            append={<EuiFormAppendPrepend {...defaultProps} />}
          />
        );

        const classes = Object.values(
          getByTestSubject(defaultProps['data-test-subj']).classList
        );

        expect(classes.some((clx) => clx.includes('compressed'))).toBe(true);
      });
    });

    describe('onClick', () => {
      it('renders a button with click handler', () => {
        const onClick = jest.fn();

        const { getByTestSubject } = render(
          <EuiFormAppendPrepend
            {...defaultProps}
            element="button"
            onClick={onClick}
          />
        );

        const element = getByTestSubject(defaultProps['data-test-subj']);

        expect(element).toBeInTheDocument();
        expect(element.tagName).toBe('BUTTON');

        fireEvent.click(element);

        expect(onClick).toHaveBeenCalledTimes(1);
      });
    });
  });
});
