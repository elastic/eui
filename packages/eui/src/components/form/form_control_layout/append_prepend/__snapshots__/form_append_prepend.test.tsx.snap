// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiFormAppend is rendered 1`] = `
<div
  aria-label="aria-label"
  class="euiFormAppend testClass1 testClass2 css-14xvn4g-side-uncompressed-euiTestCss"
  data-test-subj="test subject string"
>
  <label
    class="euiFormLabel emotion-euiFormLabel"
  >
    Label
  </label>
</div>
`;

exports[`EuiFormAppendPrepend is rendered 1`] = `
<div
  aria-label="aria-label"
  class="euiFormAppend testClass1 testClass2 css-14xvn4g-side-uncompressed-euiTestCss"
  data-test-subj="test subject string"
>
  <label
    class="euiFormLabel emotion-euiFormLabel"
  >
    Label
  </label>
</div>
`;

exports[`EuiFormPrepend is rendered 1`] = `
<div
  aria-label="aria-label"
  class="euiFormPrepend testClass1 testClass2 css-14xvn4g-side-uncompressed-euiTestCss"
  data-test-subj="test subject string"
>
  <label
    class="euiFormLabel emotion-euiFormLabel"
  >
    Label
  </label>
</div>
`;
